lib/matplotlib/_constrained_layout.py

```diff
@@ -266,397 +426,164 @@ def _make_layout_margins(ax, renderer, h_pad, w_pad):
+    trans_fig_to_panel = fig.transFigure + fig.transPanel.inverted()
+    for panel in fig.panels:
+        bbox = panel._layoutgrid.get_outer_bbox()
+        panel._redo_transform_rel_fig(
+            bbox=bbox.transformed(trans_fig_to_panel))
+        _reposition_axes(panel, renderer,
+                         w_pad=w_pad, h_pad=h_pad,
+                         wspace=wspace, hspace=hspace)
+
+    # for ax in [a for a in fig._localaxes if hasattr(a, 'get_subplotspec')]:
+    for ax in [a for a in fig.get_axes() if hasattr(a, 'get_subplotspec')]:
+
+        # grid bbox is in Figure co-ordinates, but we specify in panel
+        # co-ordinates...
+        ss = ax.get_subplotspec()
+        gs = ss.get_gridspec()
+        nrows, ncols = gs.get_geometry()
+        if gs._layoutgrid is None:
+            return
+
+        bbox = gs._layoutgrid.get_inner_bbox(rows=ss.rowspan, cols=ss.colspan)
+
+        bboxouter = gs._layoutgrid.get_outer_bbox(rows=ss.rowspan,
+                                                  cols=ss.colspan)
+
+        # transform from figure to panel for set_position:
+        newbbox = trans_fig_to_panel.transform_bbox(bbox)
+        ax._set_position(newbbox)
+
+        # move the colorbars:
+        # we need to keep track of oldw and oldh if there is more than
+        # one colorbar:
+        offset = {'left': 0, 'right': 0, 'bottom': 0, 'top': 0}
+        for nn, cbax in enumerate(ax._colorbars[::-1]):
+            if ax == cbax._colorbar_info['parents'][0]:
+                margin = _reposition_colorbar(
+                    cbax, renderer, offset=offset)
+
+
+def _reposition_colorbar(cbax, renderer, *, offset=None):
+
     """
-    Do the layout for a colorbar, to not overly pollute colorbar.py
+    Place the colorbar in its new place.
 
-    *pad* is in fraction of the original axis size.
+    Parameters
+    ----------
+    cbax : Axes for the colorbar
+
+    renderer :
+    w_pad, h_pad : float
+        width and height padding (in fraction of figure)
+    hspace, wspace : float
+        width and height padding as fraction of figure size divided by
+        number of  columns or rows
+    margin : array float
+        offset the colorbar needs to be pushed to in order to
+        account for multiple colorbars
     """
-    axlb = ax._layoutbox
-    axpos = ax._poslayoutbox
-    axsslb = ax.get_subplotspec()._layoutbox
-    lb = layoutbox.LayoutBox(
-            parent=axsslb,
-            name=axsslb.name + '.cbar',
-            artist=cax)
 
-    if location in ('left', 'right'):
-        lbpos = layoutbox.LayoutBox(
-                parent=lb,
-                name=lb.name + '.pos',
-                tightwidth=False,
-                pos=True,
-                subplot=False,
-                artist=cax)
+    parents = cbax._colorbar_info['parents']
+    gs = parents[0].get_gridspec()
+    ncols, nrows = gs.ncols, gs.nrows
+    fig = cbax.figure
+    trans_fig_to_panel = fig.transFigure + fig.transPanel.inverted()
```

Comment:
transfigure - transpanel, also pick a consistent name between pan2fig and trans_fig_to_panel :)

Reply:
This change could still be made, I think?

Reply:
I think its resolved.  I picked `trans_fig_to_panel`.  I used both at one iteration....

Reply:
It's still `fig.transPanel.inverted()`, not `- fig.transPanel`.

Reply:
Ooops!